### PR TITLE
prometheus calls should be made with explicitly generated token

### PR DESCRIPTION
**Pull request by dbasunag**
##### Short description: 
Cluster monitoring operator asks the apiserver to deploy https://github.com/openshift/cluster-monitoring-operator/blob/release-4.17/assets/prometheus-k8s/service-account.yaml. openshift-monitoring team recommends generating specific tokens and using it for prometheus query, instead of assuming imagePullSecrets would always be there. In CNV we found this field not to be reliable, specially for post upgrade scenarios.

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
    * You can add extra args to the Podman build command
        * Example: `/build-and-push-container --build-arg OPENSHIFT_PYTHON_WRAPPER_COMMIT=<commit_hash>`
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest python-module-install`: Retest python-module-install

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

File: ocp_utilities/monitoring.py [hunk 1]
```diff
@@ -48,7 +46,8 @@ def __init__(
             resource_name (str): Prometheus API resource name
             client (DynamicClient): Admin client resource
             verify_ssl (bool): Perform SSL verification on query
-            bearer_token (str): Used for query OAuth with API endpoint
+            bearer_token (str): Used for query OAuth with API endpoint, this needs to be created via oc create token command
+                example: oc create token <resource_name> -n <namespace> --duration=<number>s
```
Comment (myakove) [inline ocp_utilities/monitoring.py:50]: Can we give here a more specific example?
Maybe a command to find on which ns is Prometheus and how to generate the token.
 `oc create token prometheus-token -n <namespace> --duration=<number>s`

File: ocp_utilities/monitoring.py [hunk 2]
```diff
@@ -70,27 +69,8 @@ def _get_route(self) -> str:
     def _get_headers(self) -> Dict[str, str]:
         """Uses the Prometheus serviceaccount to get an access token for OAuth if not given"""
         LOGGER.info("Setting Prometheus headers and Obtaining OAuth token")
```
Comment (myakove) [inline ocp_utilities/monitoring.py:71]: We do not obtain the token anymore, please rephrase

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_utilities/monitoring.py (hunk 1 above)
Comment (dbasunag) [reply ocp_utilities/monitoring.py:50]: Sure. But it basically depends on the resource_name and namespace. I will add a sample that we use in cnv.

File: ocp_utilities/monitoring.py [hunk 3]
```diff
@@ -48,7 +46,10 @@ def __init__(
             resource_name (str): Prometheus API resource name
             client (DynamicClient): Admin client resource
             verify_ssl (bool): Perform SSL verification on query
-            bearer_token (str): Used for query OAuth with API endpoint
+            bearer_token (str, Required): Used for query OAuth with API endpoint, this needs to be created via oc
```
Comment (rnetser) [inline ocp_utilities/monitoring.py:49]: `bearer_token` should be set as mandatory in args definition and not as empty string

File: ocp_utilities/monitoring.py [hunk 4]
```diff
@@ -68,29 +69,9 @@ def _get_route(self) -> str:
...
     def _get_headers(self) -> Dict[str, str]:
-        """Uses the Prometheus serviceaccount to get an access token for OAuth if not given"""
-        LOGGER.info("Setting Prometheus headers and Obtaining OAuth token")
-
-        if not self.bearer_token:
-            secret = self._get_resource_secret()
-            self.bearer_token = secret.instance.metadata.annotations["openshift.io/token-secret.value"]
-
+        LOGGER.info("Setting authorization headers for prometheus call")
         return {"Authorization": f"Bearer {self.bearer_token}"}
```
Comment (rnetser) [inline ocp_utilities/monitoring.py:73]: ` self.headers` can be set directly in __init__

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_utilities/monitoring.py [hunk 5]
```diff
@@ -36,61 +34,43 @@ class Prometheus(object):
...
         """
         Args:
+            bearer_token (str, Required): Used for query OAuth with API endpoint, this needs to be created via oc
+            create token command
             namespace (str): Prometheus API resource namespace
             resource_name (str): Prometheus API resource name
             client (DynamicClient): Admin client resource
             verify_ssl (bool): Perform SSL verification on query
-            bearer_token (str): Used for query OAuth with API endpoint
+        Example to create prometheus token: oc create token prometheus-k8s -n openshift-monitoring --duration=600s
```
Comment (rnetser) [inline ocp_utilities/monitoring.py:51]: should be next to `bearer_token`

File: ocp_utilities/monitoring.py [hunk 6]
```diff
@@ -36,61 +34,43 @@ class Prometheus(object):
...
         """
         self.namespace = namespace
         self.resource_name = resource_name
         self.client = client or get_client()
         self.api_v1 = "/api/v1"
         self.verify_ssl = verify_ssl
         self.bearer_token = bearer_token
+
+        if not self.bearer_token:
+            raise ValueError("Prometheus token is required.")
```
Comment (rnetser) [inline ocp_utilities/monitoring.py:62]: why is this needed?

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by rnetser — APPROVED** [review]

File: ocp_utilities/monitoring.py (hunk 6 above)
Comment (dbasunag) [reply ocp_utilities/monitoring.py]: Removed.

File: ocp_utilities/monitoring.py (hunk 5 above)
Comment (dbasunag) [reply ocp_utilities/monitoring.py]: Done.

**Review by myakove — APPROVED** [review]
